### ISSUE TITLE
Conflict between const-related rules

### ISSUE THREAD

**Issue by grantovich** [DockYard/eslint-plugin-ember-suave#31]
Consider this code that implements an error handler in the form of a service:

```js
import Ember from 'ember';

export default Ember.Service.extend({
  setup() {
    let originalHandler = Ember.onerror;

    Ember.onerror = (error) => {
      // ... do something with the error ...

      if (originalHandler) {
        originalHandler(error);
      } else {
        throw error;
      }
    };
  }
});
```

The `let` on the first line of the function triggers `require-const-from-ember-properties`. However, changing it to a `const` instead triggers `no-const-outside-module-scope`. But the assignment needs to be in the function scope to correctly retrieve and delegate to whatever the current error handler is at the moment it's called.

I'm not sure if this has any applicability beyond the specific case of `Ember.onerror` – maybe the best option is to just disable one of the rules on that line (it's not going to keep me up at night). Still, it seems odd for two Suave-provided rules to disagree over what the correct style is. Am I overlooking something that would resolve the discrepancy?

**Comment by alexlafroscia** [comment]
Why not do:

```js
import Ember from 'ember';

const originalHandler = Ember.onerror;

export default Ember.Service.extend({
  setup() {
    Ember.onerror = (error) => {
      // ... do something with the error ...

      if (originalHandler) {
        originalHandler(error);
      } else {
        throw error;
      }
    };
  }
});
```

**Comment by alexlafroscia** [comment]
And out of curiosity, what's the use case for an error handler in a service?  I usually see this kind of code in an initializer.

**Comment by grantovich** [comment]
> out of curiosity, what's the use case for an error handler in a service?

The real code has dependencies on other services, which I believe cannot be done in an initializer.

> Why not do:

Suppose two services/initializers did this. At load time, they both grab a reference to the current `Ember.onerror` function, call it A. The first service's setup code overwrites `Ember.onerror` with its own function, B, which passes the error on to the "original" handler A. Then the second service's setup code overwrites `Ember.onerror` with its function, C, which again delegates to the "original" handler A. Unfortunately handler B has now been clobbered and will never run. The solution is to retrieve the "original" handler only when you are about to replace it.

(also suppose one of these services/initializers is located in a third-party addon, so you can't solve the problem by keeping all your `Ember.onerror` usage in one place – this is actually the case in my app)

**Comment by alexlafroscia** [comment]
> The real code has dependencies on other services, which I believe cannot be done in an initializer.

I think you should be able to do this in an instance initializer.  If you have multiple services that are playing with `Ember.onerror` and you're worried about one clobbering the other, you might want to be explicit about the order in an initializer instead.

**Comment by grantovich** [comment]
I agree being more explicit about the order would be good, but I'm not sure that would change the reason I encountered this dilemma: In order to work with third-party initializers that reassign `Ember.onerror`, my own initializer needs to store its current value ("current" as of the time it executes) so it can be delegated to.

This is really a tangent from the issue as it pertains to `ember-suave`, though 😉 My intention was less "please fix my code" (the specific code isn't even important), and more "hey I noticed these two Suave rules can contradict each other under non-contrived circumstances, what do y'all think".

**Comment by alexlafroscia** [comment]
Yeah, that's pretty interesting.  I think, when the rule was made, this kind of thing (accessing an Ember property outside of the top level of the module) wasn't a use case that was thought of. Certainly not be me, anyway.

I personally don't use `no-const-outside-module-scope` on any of the things I maintain that use Suave and just turn that rule off, because I prefer `const` for any bindings that I don't want to change.  I'm really glad you brought up the fact that those two rules potentially conflict, though.
